src/team_lib/mhba/xccl_mhba_mkeys.c

```diff
@@ -179,55 +161,123 @@ static xccl_status_t poll_umr_cq(xccl_mhba_context_t *ctx) {
+        if (status != XCCL_OK) {
+            xccl_mhba_error("create send masterkey[%d] failed",i);
+            return status;
+        }
+        status = create_master_key(ctx, &node->recv_mkey[i], node->sbgp->group_size + 1);
+        if (status != XCCL_OK) {
+            xccl_mhba_error("create recv masterkey[%d] failed",i);
+            if (mlx5dv_destroy_mkey(node->send_mkey[i])) {
+                xccl_mhba_error("mkey destroy failed (errno=%d)", errno);
+            }
+            return status;
+        }
+    }
+    status = create_and_populate_team_mkey(ctx,node,&node->team_send_mkey,node->send_mkey,0,team_size);
+    if (status != XCCL_OK) {
+        return status;
+    }
+    status = create_and_populate_team_mkey(ctx,node,&node->team_recv_mkey,node->recv_mkey,
+                                           IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_WRITE,team_size);
+    if (status != XCCL_OK) {
+        return status;
+    }
+    return XCCL_OK;
+}
+
+xccl_status_t xccl_mhba_populate_send_recv_mkeys(xccl_mhba_context_t *ctx, xccl_mhba_node_t *node,int seq_num,int block_size,int team_size){
     xccl_status_t status;
     int send_mem_access_flags = 0;
-    status = populate_mkey(ctx,node,send_mem_access_flags,node->send_mkey,node->send_umr_data);
+    int index = seq_index(seq_num);
+    status = populate_mkey(ctx,node,send_mem_access_flags,node->send_mkey[index],node->send_umr_data[index],block_size,team_size);
     if (status != XCCL_OK) {
-        xccl_mhba_error("Failed to populate send umr");
+        xccl_mhba_error("Failed to populate send umr[%d]",index);
         return status;
     }
     int recv_mem_access_flags = IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_WRITE;
-    status = populate_mkey(ctx,node,recv_mem_access_flags,node->recv_mkey,node->recv_umr_data);
+    status = populate_mkey(ctx,node,recv_mem_access_flags,node->recv_mkey[index],node->recv_umr_data[index],block_size,team_size);
     if (status != XCCL_OK) {
-        xccl_mhba_error("Failed to populate recv umr");
+        xccl_mhba_error("Failed to populate recv umr[%d]",index);
         return status;
     }
     return XCCL_OK;
 }
 
-static void update_mkey_entry(xccl_mhba_node_t *node, void *entry, struct ibv_mr *buff,const char* direction){
+static void update_mkey_entry(xccl_mhba_node_t *node, void *entry, struct ibv_mr *buff,const char* direction, xccl_mhba_coll_req_t *req){
     struct mlx5dv_mr_interleaved *mkey_entry = (struct mlx5dv_mr_interleaved *) entry;
     mkey_entry->addr = (uintptr_t) buff->addr; // TODO: Check why conversion to uintptr_t is needed and if it's correct
-    mkey_entry->bytes_count = node->block_size;
+    mkey_entry->bytes_count = req->block_size*req->args.buffer_info.len;
     mkey_entry->bytes_skip = 0;
-    mkey_entry->lkey = buff->lkey;// TODO: For "Recv Side" MasterMKey, shouldn't the entry be populated with the 'rkey' of the DR's memory region?
+    if(direction == "send") {
```

Comment:
this is the comparison of pointers. C language does not work this way. should use strcmp. Or even better, use integer flag.